File: src/data/codes/random-forest.m
clc;  % random forest
clear;
close all;

%% =========================
% SETTINGS
% ==========================
useSynthetic = true;
csvFile = 'random_forest_loan.csv';
numTrees = 100;

%% =========================
% LOAD OR CREATE DATA
% ==========================
if useSynthetic
    rng(15);
    n = 180;

    Annual_Income = 2 + 8*rand(n,1);           % lakhs
    Credit_Score = randi([450 850], n, 1);
    Employment_Years = randi([0 15], n, 1);
    Loan_Amount = 2 + 10*rand(n,1);
    Existing_Loans = randi([0 3], n, 1);

    score = 0.7*Annual_Income + 0.01*Credit_Score + 0.15*Employment_Years ...
            - 0.5*Loan_Amount - 0.8*Existing_Loans;

    Loan_Status = double(score > 7.5);

    data = table(Annual_Income, Credit_Score, Employment_Years, Loan_Amount, Existing_Loans, Loan_Status);
else
    data = readtable(csvFile);
    % Expected columns:
    % Annual_Income, Credit_Score, Employment_Years, Loan_Amount, Existing_Loans, Loan_Status
end

disp(head(data));

%% =========================
% PREPROCESSING
% ==========================
data = rmmissing(data);
data = unique(data);

X = [data.Annual_Income, data.Credit_Score, data.Employment_Years, data.Loan_Amount, data.Existing_Loans];
Y = data.Loan_Status;

%% =========================
% TRAIN-TEST SPLIT
% ==========================
cv = cvpartition(Y, 'HoldOut', 0.2);
X_train = X(training(cv), :);
Y_train = Y(training(cv), :);
X_test  = X(test(cv), :);
Y_test  = Y(test(cv), :);

%% =========================
% MODEL BUILDING
% ==========================
model = TreeBagger(numTrees, X_train, Y_train, ...
    'Method', 'classification', ...
    'OOBPrediction', 'On');

%% =========================
% PREDICTION
% ==========================
[Y_pred_cell, ~] = predict(model, X_test);
Y_pred = str2double(Y_pred_cell);

%% =========================
% EVALUATION
% ==========================
confMat = confusionmat(Y_test, Y_pred);
accuracy = sum(diag(confMat)) / sum(confMat(:));

TP = confMat(2,2); FP = confMat(1,2); FN = confMat(2,1);
precision = TP / (TP + FP + eps);
recall    = TP / (TP + FN + eps);
f1        = 2 * precision * recall / (precision + recall + eps);

disp('Confusion Matrix:');
disp(confMat);
fprintf('\nNumber of Trees = %d\n', numTrees);
fprintf('Accuracy  = %.4f\n', accuracy);
fprintf('Precision = %.4f\n', precision);
fprintf('Recall    = %.4f\n', recall);
fprintf('F1-score  = %.4f\n', f1);

%% =========================
% EFFECT OF NUMBER OF TREES
% ==========================
treeVals = [10 20 50 100 150];
accVals = zeros(size(treeVals));

for i = 1:length(treeVals)
    tempModel = TreeBagger(treeVals(i), X_train, Y_train, 'Method', 'classification');
    tempPredCell = predict(tempModel, X_test);
    tempPred = str2double(tempPredCell);
    tempConf = confusionmat(Y_test, tempPred);
    accVals(i) = sum(diag(tempConf)) / sum(tempConf(:));
end

figure;
plot(treeVals, accVals, 'o-', 'LineWidth', 1.5);
xlabel('Number of Trees');
ylabel('Accuracy');
title('Effect of Number of Trees on Random Forest Accuracy');
grid on;

figure;
confusionchart(Y_test, Y_pred);
title('Random Forest Confusion Matrix');

File: src/data/codes/random-forest.py
"""Random Forest
# pip install numpy pandas scikit-learn matplotlib
"""
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import ConfusionMatrixDisplay
import matplotlib.pyplot as plt

USE_SYNTHETIC = True
N_TREES = 100
RNG = np.random.default_rng(15)

if USE_SYNTHETIC:
    n = 180
    income = RNG.uniform(2, 10, size=n)
    credit = RNG.integers(450, 851, size=n)
    emp = RNG.integers(0, 16, size=n)
    loan_amt = RNG.uniform(2, 12, size=n)
    existing = RNG.integers(0, 4, size=n)
    score = (
        0.7 * income
        + 0.01 * credit
        + 0.15 * emp
        - 0.5 * loan_amt
        - 0.8 * existing
    )
    status = (score > 7.5).astype(int)
    data = pd.DataFrame(
        {
            "Annual_Income": income,
            "Credit_Score": credit,
            "Employment_Years": emp,
            "Loan_Amount": loan_amt,
            "Existing_Loans": existing,
            "Loan_Status": status,
        }
    )
else:
    data = pd.read_csv("random_forest_loan.csv")

print(data.head())
data = data.dropna().drop_duplicates()
cols = [
    "Annual_Income",
    "Credit_Score",
    "Employment_Years",
    "Loan_Amount",
    "Existing_Loans",
]
X = data[cols].values
y = data["Loan_Status"].values

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=42
)

model = RandomForestClassifier(n_estimators=N_TREES, random_state=42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print(f"\nN Trees = {N_TREES}")
print(f"Accuracy  = {accuracy_score(y_test, y_pred):.4f}")
print(f"Precision = {precision_score(y_test, y_pred, zero_division=0):.4f}")
print(f"Recall    = {recall_score(y_test, y_pred, zero_division=0):.4f}")
print(f"F1-score  = {f1_score(y_test, y_pred, zero_division=0):.4f}")

tree_vals = [10, 20, 50, 100, 150]
acc_vals = []
for nt in tree_vals:
    m = RandomForestClassifier(n_estimators=nt, random_state=42).fit(X_train, y_train)
    pred = m.predict(X_test)
    acc_vals.append(accuracy_score(y_test, pred))

fig, ax = plt.subplots()
ax.plot(tree_vals, acc_vals, "o-", linewidth=1.5)
ax.set_xlabel("Number of Trees")
ax.set_ylabel("Accuracy")
ax.set_title("Effect of Number of Trees on Random Forest Accuracy")
ax.grid(True)

fig2, ax2 = plt.subplots()
ConfusionMatrixDisplay.from_predictions(y_test, y_pred, ax=ax2)
ax2.set_title("Random Forest Confusion Matrix")
plt.show()
